Mathlib/Dynamics/SymbolicDynamics/Basic.lean

```diff
@@ -0,0 +1,602 @@
+    IsOpen (cylinder (A := A) (G := G) U x) := by
+  classical
+  have hopen : ∀ i ∈ (↑U : Set G), IsOpen ({x i} : Set A) := by
+    intro i _; simp
+  have hpi :
+      IsOpen (Set.pi (s := (↑U : Set G))
+                     (t := fun i => ({x i} : Set A))) :=
+    isOpen_set_pi (U.finite_toSet) hopen
+  simpa [cylinder_eq_set_pi (A := A) (G := G) U x] using hpi
+
+end CylindersOpen
+
+section CylindersClosed
+
+variable {A G : Type*} [TopologicalSpace A] [DiscreteTopology A]
+
+lemma cylinder_is_closed (U : Finset G) (x : G → A) :
+    IsClosed (cylinder (A := A) (G := G) U x) := by
+  classical
+  have hclosed : ∀ i ∈ (↑U : Set G), IsClosed ({x i} : Set A) := by
+    intro i _; simp
+  have hpi :
+      IsClosed (Set.pi (s := (↑U : Set G))
+                       (t := fun i => ({x i} : Set A))) :=
+    isClosed_set_pi hclosed
+  simpa [cylinder_eq_set_pi (A := A) (G := G) U x] using hpi
+
+end CylindersClosed
+
+/-! ## Patterns and occurrences -/
+
+section SubshiftDef
+variable (A : Type*) [TopologicalSpace A]
+variable (G : Type*) [Monoid G]
+
+/-- A subshift is a closed, shift-invariant subset. -/
+structure Subshift : Type _ where
+  /-- The underlying set of configurations. -/
+  carrier : Set (G → A)
+  /-- Closedness of `carrier`. -/
+  isClosed : IsClosed carrier
+  /-- Shift invariance of `carrier`. -/
+  shiftInvariant : ∀ g : G, ∀ x ∈ carrier, mulShift g x ∈ carrier
+
+end SubshiftDef
+
+
+section AddSubshiftDef
+variable (A : Type*) [TopologicalSpace A]
+variable (G : Type*) [AddMonoid G]
+
+/-- Additive version of the definition of subshift. -/
+structure AddSubshift : Type _ where
+  /-- The underlying set of configurations (additive group version). -/
+  carrier : Set (G → A)
+  /-- Closedness of `carrier`. -/
+  isClosed : IsClosed carrier
+  /-- Shift invariance of `carrier` for the additive shift `addShift`. -/
+  shiftInvariant : ∀ g : G, ∀ x ∈ carrier, addShift g x ∈ carrier
+
+end AddSubshiftDef
+
+attribute [to_additive existing SymbolicDynamics.Ambiant.AddSubshift]
+  SymbolicDynamics.Ambiant.Subshift
+
+/-- Example: the full shift on alphabet A. -/
+@[to_additive SymbolicDynamics.Ambiant.addFullShift]
+def fullShift (A G) [TopologicalSpace A] [Monoid G] : Subshift A G :=
+{ carrier := Set.univ,
+  isClosed := isClosed_univ,
+  shiftInvariant := by intro _ _ _; simp }
+
+attribute [inherit_doc SymbolicDynamics.Ambiant.fullShift] SymbolicDynamics.Ambiant.addFullShift
+
+/-- A finite pattern: finite support in `G` and values on it. -/
+structure Pattern (A : Type*) (G : Type*) [Monoid G] where
+  /-- Finite support of the pattern. -/
+  support : Finset G
+  /-- The value (symbol) at each point of the support. -/
+  data : support → A
+
+/-- Additive version of `Pattern`. -/
+structure AddPattern (A : Type*) (G : Type*) [AddMonoid G] : Type _ where
+  /-- Finite support of the pattern (subset of `G`). -/
+  support : Finset G
+  /-- The symbol at each point of the support. -/
+  data    : support → A
```

Comment:
```suggestion
  data : support → A
```
We avoid vertical alignment like that, because it is not robust against refactors, so creates more work for a very small gain.